### DISCUSSION TITLE
Add solid/blink cursor toggle

### DISCUSSION THREAD

**Discussion by peterkos** [helix-editor/helix#12363] [General]
It'd be great if the cursor could have a `solid` mode that does not blink.

Sublime Text has a few options("smooth", "phase", "blink", "solid"), but it seems easiest just to do `blink` and `solid`.

**Comment by the-mikedavis** [comment]
I think this is a terminal option. For example, kitty has a [`cursor_blink_interval` option](https://sw.kovidgoyal.net/kitty/conf/?highlight=blink#opt-kitty.cursor_blink_interval) that, when set to `0`, disables blinking. Which terminal emulator are you using?

**Comment by NNBnh** [comment]
"blink" and "solid" cursor can be set by using ANSI escape code lkie:

```
echo -e -n "\x1b[\x30 q" # Blinking block
echo -e -n "\x1b[\x31 q" # Blinking block
echo -e -n "\x1b[\x32 q" # Steady block
echo -e -n "\x1b[\x33 q" # Blinking underline
echo -e -n "\x1b[\x34 q" # Steady underline
echo -e -n "\x1b[\x35 q" # Blinking bar
echo -e -n "\x1b[\x36 q" # Steady bar
```

But "smooth" and "phase" is impossible due to limitation of the terminal.

Maybe this will be imperment to the GUI forntend.

**Comment by peterkos** [comment]
> I think this is a terminal option. For example, kitty has a [`cursor_blink_interval` option](https://sw.kovidgoyal.net/kitty/conf/?highlight=blink#opt-kitty.cursor_blink_interval) that, when set to `0`, disables blinking. Which terminal emulator are you using?

I use iTerm2 (3.4.15). Even with `Blinking Cursor` disabled it still blinks when resting.
(Also, curiously, the cursor shape of my iTerm setting gets overridden with the block cursor after quitting helix. Might be worth a separate bug report)

https://user-images.githubusercontent.com/2782749/162591738-48b02b21-b42d-43bb-aab6-a79a8326bfac.mp4

**Comment by fomotoshi** [comment]
I use Kitty(Mac OS) and my caret blinking is on, but when I open helix my caret blinking is off )).

**Comment by sudormrfbin** [comment]
Helix draws its own cursors since we need multiple cursors, hiding the terminal cursor.

**Comment by ChinmayMoghe** [comment]
Same issue I have faced is there a way to disable the animation ?

**Comment by eliasboegel** [comment]
I would like this too. The block cursor for me is indeed solid and non-flashing, but the bar cursor flashes (even when moving the cursor), making it hard to see its position sometimes when moving the cursor around.

**Comment by Tubbles** [comment]
> (Also, curiously, the cursor shape of my iTerm setting gets overridden with the block cursor after quitting helix. Might be worth a separate bug report)

Noticed this as well so created a ticket #5364

**Comment by 7ombie** [comment]
My Helix cursor never blinks (including at the command line, in the file picker etc - where the terminal cursor's *color* is reused by Helix). The cursor blinks correctly in zsh, outside of Helix. I'm using iTerm on macOS.

A couple of other people have reported the same issue (@ChzenChzen and @eliasboegel, also on macOS).

**Comment by 7ombie** [comment]
***There are now two bugs and a feature request in this issue.***

There is a bug with cursors not blinking, another bug where the cursor is not properly reset when Helix exits, as well as a feature request for new cursor configuration options (re. blinking).

**Comment by Alacho2** [comment]
Just trying out Kitty and I came to say that the cursor is not blinking in Helix but blinking just fine in Vim. :)

**Comment by JonnieCache** [comment]
Also seeing a non-blinking cursor on ubuntu 24.04 using the builtin terminal. The cursor blinks correctly outside of helix, in the shell and in vim for example.

**Comment by hholst80** [comment]
> _**There are now two bugs and a feature request in this issue.**_
> 
> There is a bug with cursors not blinking, another bug where the cursor is not properly reset when Helix exits, as well as a feature request for new cursor configuration options (re. blinking).

This is a workaround for restoring blinking (if this is what you have) after closing helix:

1) rename helix binary to `hx.bin` or put a executable called `hx` in your PATH *before* the actual hx binary. I have installed helix in /usr/local/bin because I made some patching of it.
2) put this wrapper script in your path. remember to make it executable by setting the execute bit (chmod 755 hx).

```
#!/bin/sh

hx.bin "$@"           # exec cannot be used because we need to run a cleanup echo after the call
echo -n -e "\e[?12h"  # Enable cursor blinking
```

Note that Neovim seems to have this exact issue too, so it could be reasonable to handle the fallback in the shell prompt.

**Comment by Kajiih** [comment]
On mac, I tried on Ghostty, Iterm2 and the built-in terminal, and I couldn't have the block cursor blink in any of them, but interestingly, in Iterm2, the line cursor is blinking.
